File: python/5kyu/directions_reduction.py
# https://www.codewars.com/kata/550f22f4d758534c1100025a

def dirReduc(arr):
    dir1 = ' '.join(arr)
    dir2 = dir1.replace('NORTH SOUTH', '').replace('SOUTH NORTH', '').replace('EAST WEST', '').replace('WEST EAST', '')
    dir3 = dir2.split()
    return dirReduc(dir3) if len(dir3) < len(arr) else dir3


print(dirReduc(["NORTH", "SOUTH", "SOUTH", "EAST", "WEST", "NORTH", "WEST"]), ["WEST"])

File: go/5kyu/directions_reduction.go
// https://www.codewars.com/kata/550f22f4d758534c1100025a

package main

import (
	"fmt"
)

func DirReduc(arr []string) []string {
	opposites := map[string]string{
		"NORTH": "SOUTH",
		"SOUTH": "NORTH",
		"EAST":  "WEST",
		"WEST":  "EAST",
	}

	var reduced []string
	for _, dir := range arr {
		if len(reduced) > 0 && reduced[len(reduced)-1] == opposites[dir] {
			reduced = reduced[:len(reduced)-1]
		} else {
			reduced = append(reduced, dir)
		}
	}

	return reduced
}

func main() {
	fmt.Println(DirReduc([]string{"NORTH", "SOUTH", "SOUTH", "EAST", "WEST", "NORTH", "WEST"}), []string{"WEST"})
}
